test/index.spec.js

```diff
@@ -1719,3 +1745,276 @@ describe('api route file analysis', () => {
     )
   })
 })
+
+const middlewareSourceTs = /* typescript */ `
+import { NextResponse } from 'next/server'
+export async function middleware(req: NextRequest) {
+  return NextResponse.next()
+}
+`
+
+const middlewareSourceJs = /* javascript */ `
+import { NextResponse } from 'next/server'
+export async function middleware(req) {
+  return NextResponse.next()
+}
+`
+
+const wait = (seconds = 0.5) => new Promise((resolve) => setTimeout(resolve, seconds * 1000))
+
+const middlewareExists = () => existsSync(resolve('.netlify', 'middleware.js'))
+
+describe('onPreDev', () => {
+  let runtime
+  beforeAll(async () => {
+    runtime = await nextRuntimeFactory({}, { events: new Set(['onPreDev']) })
+  })
+
+  it('should generate the runtime with onPreDev', () => {
+    expect(runtime).toHaveProperty('onPreDev')
+  })
+
+  it('should compile middleware', async () => {
+    await moveNextDist('.next', true)
+    await writeFile(path.join(process.cwd(), 'middleware.ts'), middlewareSourceTs)
+    expect(middlewareExists()).toBeFalsy()
+
+    await runtime.onPreDev(defaultArgs)
+    await wait()
+
+    expect(middlewareExists()).toBeTruthy()
+  })
+})
+
+fdescribe('the dev middleware watcher', () => {
+  const watchers = []
+
+  afterEach(async () => {
+    await Promise.all(
+      watchers.map((watcher) => {
+        console.log('closing watcher')
+        return watcher.close()
+      }),
+    )
+    watchers.length = 0
```

Comment:
Yes, the watcher carry on until `close()` has been called, even if there are no events registered on the watcher. I've done it outside of the tests themselves is because otherwise if a test fails the cleanup code won't run. By pushing them onto an array I can have multiple watchers and be sure they'll all be cleaned up at the end, even if the test fails.